### PR TITLE
App deletion cleanup

### PR DESCRIPTION
**Pull request by lmars**
Opening as an early PR for feedback.

Things that need to be done:

- [x] remove cached git repo (and implement test)
- [x] move cleanup from client into background worker

I would like some feedback on the following:

Once the cleanup is moved to the background, the client will need to wait for the job to finish so implies some kind of event stream, but I don't think adding yet another `xxx_events` db table with an SSE endpoint is the way forward. Ideally there would be a single "app events" stream including formation changes, deployments, job events and now app deletion events. The question then is should these events go into a single db table, or should they go to the logaggregator and clients stream from there instead?

/cc @titanous

### CODE REVIEW

**Comment by titanous** [comment]
> Ideally there would be a single "app events" stream including formation changes, deployments, job events and now app deletion events. The question then is should these events go into a single db table, or should they go to the logaggregator and clients stream from there instead?

A DB table makes sense. The log aggregator does not provide durability guarantees.

File: controller/client/client.go [hunk 1]
```diff
@@ -190,8 +190,37 @@ func (c *Client) UpdateApp(app *ct.App) error {
...
+	}
+	a.DeletedRoutes = routes
+
+	// remove resources
+	resources, err := c.AppResourceList(appID)
+	if err != nil {
+		return nil, err
+	}
+	for _, resource := range resources {
+		if err := c.DeleteResource(resource.ProviderID, resource.ID); err != nil {
```
Comment (titanous) [inline controller/client/client.go]: Resources can be associated with multiple apps, only makes sense to destroy if this is the last associated app.

**Comment by lmars** [comment]
@titanous comments addressed

File: controller/worker/app_deletion/app_deletion.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		if err != nil {
+			log.Error("error creating app cache delete request", "err", err)
+			return err
+		}
+		res, err := http.DefaultClient.Do(req)
+		if err != nil {
+			log.Error("error performing app cache delete request", "err", err)
+			return err
+		}
+		res.Body.Close()
```
Comment (titanous) [inline controller/worker/app_deletion/app_deletion.go]: Should check that the status code is 200

**Comment by lmars** [comment]
@titanous comments addressed

**Comment by titanous** [comment]
LGTM

**Pull request merged**